Issue 23078 unable to add categories to bundles

freddyucv opened the pull request:
### Proposed Changes
* When we add a category into the Bundle before this change we added the 'CAT' word into the bundle and later when the bundle is genereted it added all the Top Categories into the bundle
* Now when we add the Category the Category's inode is added and later all the other top categories are added as dependencies.

https://github.com/dotCMS/core/compare/issue-23078-Unable-to-add-categories-to-bundles?expand=1#diff-45822bf92f039131f8acdcd3294b73b410249dec48c45cba2bd88cceff6a116eL1014-R1060

* When a Category is added directly by the user into the bundle then We add it into the manifest with right reason

https://github.com/dotCMS/core/compare/issue-23078-Unable-to-add-categories-to-bundles?expand=1#diff-292aa9b2b6ead8a5061b3ad0b4465cced09f4ba4771217cec2d5845a172e1ddfR314

* When the others categories are added, they are added as Dependency into the Manifest

https://github.com/dotCMS/enterprise/pull/890/files

freddyDOTCMS committed:
#23078 Add catgory into the Bundle

freddyDOTCMS committed:
Running with custom ee version

wezell commented:
@freddyucv  My understanding was that when we pushed categories, we pushed all categories. This is because a child category is dependent on its parents.  Are we still doing this?  Or does this recurse up the category tree adding the parent categories of the selected categories?  Or is it just pushing the the top level (which would be wrong)

freddyDOTCMS committed:
Fixing test

freddyucv commented:
> @freddyucv My understanding was that when we pushed categories, we pushed all categories. This is because a child category is dependent on its parents. Are we still doing this? Or does this recurse up the category tree adding the parent categories of the selected categories? Or is it just pushing the the top level (which would be wrong)

Yes, we are doing the same, the different now is that we are saving the category what we added into the publishing_queue table so now we are not getting a empty bundle when we list it into the publishing queque portlet.

But later we added all the Categories when the bundle is build.

The other different now is that we are registering all in the manifest file, so the category add directly is mark as "Add directly by the user", and the others are mark as "Dependency from: [The one add by the user]", and the children are mark as dependency from its parent

dsilvam commented:
We need to change the option "Add to Bundle" to something generic like "Add All To Bundle", and then in the bundle we should not show any cat in particular but a general text like "Syncing All Categories"

freddyDOTCMS committed:
#23078 Refactoring

freddyDOTCMS committed:
reverse

dsilvam commented on dotCMS/src/main/webapp/WEB-INF/messages/Language.properties:
```diff
@@ -117,6 +117,7 @@ add-tag = Add Tag
 Add-Template = Add Template
 add-template = Add Template
 Add-To-Bundle = Add To Bundle
+Add-All-Categories-To-Bundle = Add all categories to bundle
```
Capital C and B. "Add all Categories to Bundle"

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/f391675dc6195856bc48cfd178fd3253b1177121

freddyDOTCMS committed:
Change to capital letters

freddyDOTCMS committed:
Fixing test

jdotcms commented on dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java:
```diff
@@ -117,14 +120,22 @@ public static File getFolderFilePath(final Folder folder, final File bundleRoot)
             throws DotSecurityException, DotDataException {
 
+        final String bundleId = bundleRoot.getAbsolutePath()
+                .substring(bundleRoot.getPath().lastIndexOf("/") + 1);
```
Use StringPool.FORWARD_SLASH

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-17175f0a859646e450f675099713c3f0e1d2817928a7ebdee304463353a5ddccR125

jdotcms commented on dotCMS/src/integration-test/java/com/dotcms/enterprise/publishing/remote/bundler/FileBundlerTestUtil.java:
```diff
@@ -117,14 +120,22 @@ public static File getFolderFilePath(final Folder folder, final File bundleRoot)
                 + APILocator.getHostAPI().findSystemHost().getHostname()
                 + File.separator
-                + category.getInode() + ".category.dpc.xml";
+                + category.getInode() + (categoryAddDirectly ? ".category.xml" : ".category.dpc.xml");
```
these should be constants

freddyucv replied:
this is a test class so I think it is ok to be hardcoding

jdotcms commented on dotCMS/src/integration-test/java/com/dotcms/publishing/manifest/CSVManifestBuilderTest.java:
```diff
@@ -181,7 +181,7 @@ private static TestCase getFolderTestCase() {
         final Category category = new CategoryDataGen().nextPersisted();
         final String line = list(PusheableAsset.CATEGORY.getType(), category.getInode(), StringPool.BLANK,
-                 category.getTitle(), "", "")
+                 category.getCategoryName(), "", "")
```
StringPool.Blank

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-d2fbe7006d2de513acebc78bd7b762e626273a7ac2c7e0471217bf6785c71afcR184

jdotcms commented on dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:
```diff
@@ -83,6 +84,8 @@ private static Map<String, Object> getMap(final String id, final String type,
             result = getMapForContentlet(id);
         } else if (isLanguage(type)) {
             result = getMapForLanguage(id);
+        } else if ("CAT".equals(id)) {
```
this should be a constant

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/49a6bcf7c5d48d87d283555e1283b87a17e75169

jdotcms commented on dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:
```diff
@@ -113,6 +116,10 @@ private static Map<String, Object> getMap(final String id, final String type,
     }
 
+    private static boolean isCategory(final String type) {
+        return type.equals(PusheableAsset.CATEGORY.getType());
```
probably the constant should be at the left side such as 

PusheableAsset.CATEGORY.getType().equals(type)

freddyucv replied:
really I dont need this method anymore https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-3389c7193d609b29dbc719b8c559e5149211bb7b105aa9bee2eb84b1c7f733b1L119

jdotcms commented on dotCMS/src/main/java/com/dotcms/publisher/business/PublishQueueElementTransformer.java:
```diff
@@ -126,6 +133,10 @@ private static Map<String, Object> getMapForLanguage(final String id) {
     }
 
+    private static Map<String, Object> getMapForCategory(){
+        return map(TITLE_KEY, "Syncing All Categories");
```
i18n?

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-3389c7193d609b29dbc719b8c559e5149211bb7b105aa9bee2eb84b1c7f733b1R136

jdotcms commented on dotCMS/src/main/java/com/dotcms/publisher/pusher/PushPublisherConfig.java:
```diff
@@ -134,6 +134,27 @@ public boolean justIncludesUsers() {
+			return false;
+		}
+
+		for (PublishQueueElement asset : this.getAssets()) {
```
set to final

jdotcms replied:
you may reduce this to return !this.getAssets().stream.anyMatch (asset -> !"CAT".equalsIgnoreCase(asset.getAsset()))

freddyucv replied:
https://github.com/dotCMS/core/pull/23183/commits/efcc202067d1ebc9384341b6869f6b78aa350bc7#diff-28ca72a88db7305591d8dac2158e7eb1c45b87e08805666f31c0f0be5b4089cfR149

jdotcms commented on dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java:
```diff
@@ -311,6 +313,12 @@ public void setDependencies()
+			} else if ("CAT".equals(asset.getAsset())) {
+				config.writeIncludeManifestItem((ManifestItem) () -> new ManifestInfoBuilder()
+						.objectType(PusheableAsset.CATEGORY.getType())
+						.title("Syncing All Categories")
```
i18n

freddyucv replied:
done https://github.com/dotCMS/core/pull/23183/commits/49a6bcf7c5d48d87d283555e1283b87a17e75169#diff-292aa9b2b6ead8a5061b3ad0b4465cced09f4ba4771217cec2d5845a172e1ddfR321

freddyDOTCMS committed:
Refactoring

freddyDOTCMS committed:
more refactoring
